AssignmentNotUsedInspection false positives - incorrect variable scoping and others

BZngr opened the pull request:
Closes #4913 
Closes #4754 

1. Removes 'block-based' evaluations.
2. Adds support for `static` variables and variables within static members.
3. Correctly handles prior assignments referenced by a subsequent assignment, e.g.,
```
Function Test() As String
    Dim fizz As String
    fizz = "hello" '<--- this assignment previously flagged as unused'
    fizz = fizz & " world"
    Test = fizz
End Function
```

MDoerner commented:
Thinking a bit more about the `Resume` case, could you please add a test with a single overall error handler, whose OnErrorStatement is at the start of the method. Then there should be two assignments, one used and one not used. This should still return a result. I think, the way the code is set upright now, it will not.

BZngr commented on Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs:
```diff
@@ -58,54 +73,101 @@ protected override IEnumerable<IdentifierReference> ReferencesInModule(Qualified
+                        ? FindUnusedAssignmentNodes(tree, localVariable, allAssignmentsAndReferences)
+                        : allAssignmentsAndReferences.OfType<AssignmentNode>();
+
+            return unusedAssignmentNodes.Except(FindDescendantsOfNeverFlagNodeTypes(unusedAssignmentNodes))
```
Yes - clearer and results in a few less lines of code.

BZngr commented on Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs:
```diff
@@ -58,54 +73,101 @@ protected override IEnumerable<IdentifierReference> ReferencesInModule(Qualified
         }
 
-        protected override bool IsResultReference(IdentifierReference reference, DeclarationFinder finder)
+        private static IEnumerable<AssignmentNode> FindDescendantsOfNeverFlagNodeTypes(IEnumerable<AssignmentNode> flaggedAssignments)
```
This method is now deleted and replaced by `IsDescendentOfNeverFlagNode`...though it definitely had the issues you pointed out.

BZngr commented on Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs:
```diff
@@ -131,31 +193,21 @@ private static bool IsPotentiallyUsedViaJump(IdentifierReference resultCandidate
+        private static bool TryGetRelevantJumpContext<T>(IdentifierReference resultCandidate, out T ctxt) where T : ParserRuleContext
         {
             ctxt = resultCandidate.ParentScoping.Context.GetDescendents<T>()
                                     .Where(sc => sc.Start.Line > resultCandidate.Context.Start.Line
```
I have changed it to use `Selection` for comparisons but had to modify the suggested code slightly:
```
            ctxt = resultCandidate.ParentScoping.Context.GetDescendents<T>()
                                    .Where(descendent => descendent.GetSelection() > resultCandidate.Selection)
                                    .OrderBy(descendent => descendent.GetSelection())
                                    .FirstOrDefault();

```

BZngr commented:
@MDoerner First, thanks for the awesome review.  I have updated all the code sections per your comments.  If I did not directly reply to the code comment - you can assume the answer is 'Done'.

Regarding the initial general comments:
1. Line 231 concern: changed to a `switch` statement as recommended.
2. Line numbers: yes - original implementation missed the mark entirely.  The correction was pretty straight forward though.  Thanks for spotting this.
3. `Resume` is considered in the analysis to reduce false positives from used assignments within the error handler section.
4.  The remaining points are valid as well - and, are left as a conceded analysis gaps until they can be addressed using improved code path analysis.

I added the following test per your last comment above:
```
        [Test]
        public void ResumeStmtTest_FromComments()
        {
            string code =
$@"
Public Function Inverse(value As Double) As Double
On Error GoTo ErrorHandler:
    Dim ratio As Double
    ratio = 0# '<== unused
    ratio = 1# / value '<== used
    Inverse = ratio
    Exit Function
ErrorHandler: 
    ratio = 0#  '<== possibly used
    Resume
End Function
";
            var results = InspectionResultsForStandardModule(code);
            Assert.AreEqual(1, results.Count());
        }

```

The test passed without changes to code.  So, if this is the test you had in mind, I do not intend to add it to the code base.
